icefoganalytics/internal-data-portal
# Part 3: Data Owner: Edit an Existing Dataset - Add Keywords and Owner Notes

Fixes https://github.com/icefoganalytics/internal-data-portal/issues/28

Relates to:
- https://github.com/icefoganalytics/internal-data-portal/issues/7
- [0001, Data Modeling, 2024-01-05](https://docs.google.com/document/d/1Mdbl_Cy0BgKiq0DGdvOKUqJn-FdgbF0LSVykwf__FY8/edit?pli=1)
- [0002, UI Modeling, 2024-01-08](https://docs.google.com/document/d/1tP63Ob4XPBGlASPHVf1GnNb3hlIkybmWBqJDA61u3J0/edit)

# Context

## Dataset Description Manage Page - /dataset/:slug/description/manage

The edit version of the pages doesn’t seem to exist in the wireframes, so I would need another layer of User Role components to handle the different viewer types if I build this out. It might make more sense to just use auto-save for everything.

### Front-end routes

Various tabs on the same level as this component.

### Back-end routes

* /datasets/:id - GET 
* /datasets/:id - PATCH, update data set
* /tags - GET, list tags
* /tags - POST, create new tags
* /taggings - GET, list taggings
* /taggings - POST, associate existing tag with dataset
* /taggings/:id - DELETE, disassociate tag with data, orphaned tags should cleaned up
* /tags - POST, create new tag prior to associating it with a dataset via a tagging
* /users/search?email=...

### Components

* `BaseLayout.vue` - see previous definition
* `DatasetLayout.vue`
* `DatasetDescriptionManagePage.vue` - for data owner, sys admin and business analyst
    * `SubjectFormCard.vue`
        * Keywords - TagsCombobox.vue, multi select for tags probably needs a lot of very complex wiring to support mapping tags through taggings to dataset by id, and also supporting creation of new tags and linking them via taggings
    * `OwnerNotesFormCard.vue`
        * Text input field - Dataset#owner_notes

### Source Wireframe

![image](https://github.com/icefoganalytics/internal-data-portal/assets/23045206/68e8cbac-e574-4f8d-a048-631d877e8ecc)

# Implementation

Add tagging wiring that save automatically.
Add auto-save to the primary form as well and the owner notes for continuity.
Tags are created as needed, and destroyed when all references are lost.
Note: Opted not to set up autosave for "Owner" panel.

# Screenshots

[Screencast from 2024-02-13 12-28-32.webm](https://github.com/icefoganalytics/internal-data-portal/assets/23045206/37a48145-6d76-4225-a6e5-b8a9276126d2)

# Testing Instructions

1. Run the test suite via `dev test` (or `dev test_api`)
2. Boot the app via `dev up`
3. Log in to the app at http://localhost:8080
4. Create a dataset if you don't have one from the /dashboard page.
5. After creation you will be redirected to the manage page for that dataset.
6. Check that all the fields of the "Data Description" form auto-save.
7. Check that the Subject keywords auto-save.
8. Check that the owner notes auto-save.
